router/queryrange.go

```diff
@@ -119,30 +125,46 @@ func parseLegacyRequest(req *pbsearch.RouterRequest, head uint64, lib uint64, he
 
 	if req.Descending {
 		if req.StartBlock == 0 {
-			// FIXME: why do we add StartBLock when it should be 0
-			highBlkNum = virtualHead + int64(req.StartBlock)
-			if highBlkNum < 1 {
-				return nil, derr.Statusf(codes.InvalidArgument, "invalid start block: head or lib - %d goes beyond first block (head: %d, lib: %d)", -highBlkNum, head, lib)
+			highBlkNum = virtualHead
+			if highBlkNum < int64(absoluteTruncationLowBlockNum) {
+				return nil, derr.Statusf(codes.InvalidArgument, "invalid start block: (head or lib - %d) is lower than the lowest block served by this endpoint [%d] (head: %d, lib: %d)", req.StartBlock, absoluteTruncationLowBlockNum, head, lib)
 			}
 		} else {
 			highBlkNum = int64(req.StartBlock)
+			if req.StartBlock < absoluteTruncationLowBlockNum {
+				return nil, derr.Statusf(codes.InvalidArgument, "invalid start block: %d is lower than the lowest block served by this endpoint [%d]", req.StartBlock, absoluteTruncationLowBlockNum)
+			}
+
 		}
 
 		lowBlkNum = highBlkNum - int64(req.BlockCount)
-		if lowBlkNum <= 0 { // we tolerate super high block_count
+		if lowBlkNum <= int64(absoluteTruncationLowBlockNum) { // we tolerate super high block_count
+			lowBlkNum = int64(absoluteTruncationLowBlockNum)
+		}
+		if lowBlkNum < 1 {
 			lowBlkNum = 1
 		}
 	} else { // ASCENDING
 		switch {
 		case req.StartBlock == 0:
 			lowBlkNum = 1 // original behaviour, contrary to what the docs say. will not affect block count (0->100 == 1->100)
+			if absoluteTruncationLowBlockNum != 0 {
+				lowBlkNum = int64(absoluteTruncationLowBlockNum)
+			}
 		case req.StartBlock < 0:
 			lowBlkNum = virtualHead + int64(req.StartBlock)
 
 			if lowBlkNum < 1 {
-				return nil, derr.Statusf(codes.InvalidArgument, "invalid start block: head or lib - %d goes beyond first block (head: %d, lib: %d)", lowBlkNum, head, lib)
+				return nil, derr.Statusf(codes.InvalidArgument, "invalid start block: (head or lib - %d) is lower than first block (head: %d, lib: %d)", req.StartBlock, head, lib)
```

Comment:
This will also be weird always yielding `(head or lib - -X)` since `req.StartBlock` is necessary negative in this `if` branch